README.md
# OpenBCI_Ganglion_Libraries
Firmware Libraries for the OpenBCI Ganglion Board


changelog.md
# v2.0.1

### Bug Fixes

* Adjust accelerometer axes to match Cyton #14

# v2.0.0

### New Features

* Added Wifi Shield Compatibility
* Added ability to change sample rates

### Breaking Changes

* Changed MCP3912 setup to NOT accept sample rate passed into it

## Release Candidate 2

### Bug Fixes

* LED light could have held in off state if connected to wifi shield while LED is off.

### Enhancements

* Accelerometer enabled by default with wifi.

## Release Candidate 1

Initial release candidate

## Beta4

### Enhancements

* Needed to update sample rate setting functions to match the Cyton responses and thus work with wifi drivers.
* Add more helper printing functions to reduce code string foot prints such as `::printSampleRate()`, `::printFailure()`, and `::printSuccess()`

## Beta3

### New Features

* Send gains after connecting to Wifi shield

## Beta2

The overall goal was to clean the wifi code out of the library so it would not be needed when you are building a bare board.

### Bug Fixes

* Weird timing issue with wifi shield

### Breaking Changes

* Removed all wifi code and put into [new library](https://github.com/OpenBCI/OpenBCI_Wifi_Master_Library) that must be included! The new library is a called [OpenBCI_Wifi_Master_Library](https://github.com/OpenBCI/OpenBCI_Wifi_Master_Library). It is simply included when wifi is wanted.
* Removed `.loop()` function from library and all other `wifi*.()` functions.
* `DefaultGanglion.ino` now has wifi code.

### Files

* Removed `WifiGanglion.ino`
* Add `GanglionNoWifi.ino` example

## Beta1

* Initial release with wifi

# v1.1.2

### Bug fixes

* Changed the BLE on receive to use a ring buffer. See #2

# v1.1.1

* Changed the way SD file names are generated to allow for OTA programming.

# v1.0.0

* Initial Release


Honor A Loved One.cpp
//==========================================================//
//     OpenBCI Kickstarter Pledge - Honor A Loved One       //
//==========================================================//

/*
Name Of Loved One To Be Honored -- Note About Honored Loved One
C.R. Zimmerman & Family -- all the best
Leonidas Hofmann -- Always bravely carrying on despite Tourette's and OCD. You rock!
Silvia Escamilla -- Te admiro, por tu voluntad y por tu fuerza, por tu cariño y por tu alegría.
Jillian Burns -- Gone to Young, Forever in our hearts.
Grandma Wang -- My grandmother suffered a stroke in 2005 and this is dedicated to her.
Kyle Scott Dickinson -- My awesome autistic brother, I'm also his personal care attendant at the north shore arc. -- He’s a great brother in fact he won't cry from pain yet only when I'm upset lol.
Robert S. Patton (my Dad) -- My Dad's life was ended by Alzheimer's. Fortunately for him, the disease took its toll quickly, less than 5 years. For him, and so many others I know, knew, and will know I encourage the founders to continue their work.
Ruth Turner -- You were always loved and are missed
Mark Wiegand -- At some point in time and space, we make you walk again.
Eleonora Pawlowska -- In memory of my beloved mother. I wish she was still here with all of us.
Dr Frederick Fries -- RIP 2010. Love you dad.
Leibniz -- Nothing is accomplished all at once, and it is one of my great maxims, and one of the most completely verified, that Nature makes no leaps: a maxim which I have called the law of continuity.
Ubirakitan Maciel -- A person dedicated to make Neuroscience accessible to people
A. Sterling -- My father survived a brain aneurysm 16 years ago He just turned 90!
Broome Spiro -- Different in an excellent sort of way.
Lauren S. -- Your depression doesn't need to define you
Stephanie Y. -- Hope you found your dream.
My dad -- Essential tremor
Peter Beals -- artist, the kindest man I ever know
Jacqueline --- RIP
*/


Definitions_Ganglion.h
//
//

#ifndef _Definitions_Ganglion_h
#define _Definitions_Ganglion_h

#define LED 23
#define LED_DELAY_TIME 300
#define RING_SIZE 101
#define SERIAL_BUFFER_LENGTH  20
#define MAX_BYTES_PER_PACKET  19

// Byte IDs
#define ID_RAW_DATA 0
#define ID_SAMPLE_18_MIN 1
#define ID_SAMPLE_18_MAX 100
#define ID_SAMPLE_19_MIN 101
#define ID_SAMPLE_19_MAX 200
#define ID_Z_1  201
#define ID_Z_2  202
#define ID_Z_3  203
#define ID_Z_4  204
#define ID_Z_R  205
#define ID_MULTI_PACKET 206
#define ID_MULTI_PACKET_STOP  207

// Serial Command Characters
#define DEACTIVATE_CHANNEL_1 '1'
#define DEACTIVATE_CHANNEL_2 '2'
#define DEACTIVATE_CHANNEL_3 '3'
#define DEACTIVATE_CHANNEL_4 '4'
#define ACTIVATE_CHANNEL_1 '!'
#define ACTIVATE_CHANNEL_2 '@'
#define ACTIVATE_CHANNEL_3 '#'
#define ACTIVATE_CHANNEL_4 '$'
#define START_DATA_STREAM 'b'
#define STOP_DATA_STREAM 's'
#define ENABLE_ACCELEROMETER 'n'
#define DISABLE_ACCELEROMETER 'N'
#define ENABLE_SYNTHETIC_DATA '['
#define DISABLE_SYNTHETIC_DATA ']'
#define SOFT_RESET 'v'
#define REPORT_REGISTER_SETTINGS '?'
#define ENABLE_OTA '>'
#define OPENBCI_SAMPLE_RATE_SET '~'

#define OPENBCI_CHANNEL_1 '1'
#define OPENBCI_CHANNEL_2 '2'
#define OPENBCI_CHANNEL_3 '3'
#define OPENBCI_CHANNEL_4 '4'
#define OPENBCI_CHANNEL_REF '5'

#define OPENBCI_Z_TEST_START 'z'
#define OPENBCI_Z_TEST_STOP 'Z'

//  MCP
#define DEACTIVATE 0
#define ACTIVATE 1
#define ENABLE_0   0x000E0000
#define ENABLE_1   0x000D0000
#define ENABLE_2   0x000B0000
#define ENABLE_3   0x00070000
#define DISABLE_0  0x00010000
#define DISABLE_1  0x00020000
#define DISABLE_2  0x00040000
#define DISABLE_3  0x00080000

#define MCP_SS  7   // MCP Slave Select on Simblee pin 13
#define MCP_DRDY 8   // MCP DataReady on Simblee pin 8
#define MCP_RST 20   // MCP Reset: active low

#define MCP_ADD  0x40
#define MCP_READ  0x01
#define MCP_WRITE  0x00

#define DEV_ADD  0x40

#define CHAN_0  0x00       //00   |        | -GROUP
#define CHAN_1  0x02       //01   |-TYPE  _|
#define CHAN_2  0x04       //02   |        | -GROUP
#define CHAN_3  0x06       //03  _|       _|

#define MOD_VAL  0x10      //08   |        |
#define PHASE  0x14        //0A   |        |-GROUP
#define GAIN  0x16         //0B   |       _|
#define STATUSCOM  0x18    //0C   |        |
#define CONFIG_0  0x1A     //0D   |        |-GROUP
#define CONFIG_1  0x1C     //0E   |       _|
#define OFFCAL_0  0x1E     //0F   |-TYPE   | -GROUP
#define GAINCAL_0  0x20    //10   |       _|
#define OFFCAL_1  0x22     //11   |        | -GROUP
#define GAINCAL_1  0x24    //12   |       _|
#define OFFCAL_2  0x26     //13   |        | -GROUP
#define GAINCAL_2  0x28    //14   |       _|
#define OFFCAL_3  0x2A     //15   |        | -GROUP
#define GAINCAL_3  0x2C    //16   |       _|
#define LOK_CRC  0x3E      //1F  _|        >-GROUP

#define GAIN_1  0x00000000
#define GAIN_2  0x00000249
#define GAIN_4  0x00000492
#define GAIN_8  0x000006CB
#define GAIN_16 0x00000924
#define GAIN_32 0x00000B6D

//  0x0038E050
// #define SAMPLE_25   0x00030000
// #define SAMPLE_50   0x00020000
// #define SAMPLE_100  0x00010000
// #define SAMPLE_200  0x00000000


//#define OPENBCI_NUMBER_OF_BYTES_SETTINGS_Z_TEST 3
#define HALF_PERIOD 3000
#define HALF_WAVE 100
#define SHUNT 100
#define SHUNT_SENSOR_GAIN 1000
#define UA_SAMPLE_TIME 500
#define UA_SAMPLE_LIMIT 1000


/*
 *  AD5621
 */
#define DAC_MASK  0x3FFC
#define SHUNT_SENSOR 6     // reads the current shunt sensor
#define DAC_SS    19       // DAC slave select
#define Z_TEST_1  30       // pin control for Z test channel 1
#define Z_TEST_2  29       // pin control for Z test channel 2
#define Z_TEST_3  28       // pin control for Z test channel 3
#define Z_TEST_4  25       // pin control for Z test channel 4
#define Z_TEST_REF  22     // pin control for Z test channel REF
// CONTROL BYTES
#define DAC_EN    0x0000  // enables setting the DAC output 12 bit resolution
#define DAC_1K    0x7000  // disables DAC with output tied 1K to GND
#define DAC_100K  0x8000  // disables DAC with output tied 100K to GND
#define DAC_Z     0xC000  // disables DAC with output in high Z
#define DAC_MAX   0xFFFF  // highest possible value of DAC = 4095




/*
 * LIS2DH
 */

#define LIS2DH_SS   10
#define LIS_DRDY    13

#define READ_REG    0x80
#define READ_MULTI    0x40

#define LIS3DH_MODE   3 // c pol =1, c pha = 1, mode = 3

#define STATUS_REG_AUX  0x07
#define OUT_TEMP_L      0x0C
#define OUT_TEMP_H      0x0D
#define INT_COUNTER_REG 0x0E
#define WHO_AM_I        0x0F
#define TEMP_CFG_REG    0x1F
#define CTRL_REG1       0x20
#define CTRL_REG2       0x21
#define CTRL_REG3       0x22
#define CTRL_REG4       0x23
#define CTRL_REG5       0x24
#define CTRL_REG6       0x25
#define REFERENCE       0x26
#define STATUS_REG2     0x27
#define OUT_X_L         0x28
#define OUT_X_H         0x29
#define OUT_Y_L         0x2A
#define OUT_Y_H         0x2B
#define OUT_Z_L         0x2C
#define OUT_Z_H         0x2D
#define FIFO_CTRL_REG   0x2E
#define FIFO_SRC_REG    0x2F
#define INT1_CFG        0x30
#define INT1_SOURCE     0x31
#define INT1_THS        0x32
#define INT1_DURATION   0x33
#define INT2_CFG        0x34
#define INT2_SOURCE     0x35
#define INT2_THS        0x36
#define INT2_DURATION   0x37
#define CLICK_CFG       0x38
#define CLICK_SRC       0x39
#define CLICK_THS       0x3A
#define TIME_LIMIT      0x3B
#define TIME_LATENCY    0x3C
#define TIME_WINDOW     0x3D
#define ACT_THS         0x3E
#define ACT_DUR         0x3F

/*
* SD CARD
*/
#define SD_SS 17

/** Wifi Stuff */
#define NUM_CHANNELS 4
#define OPENBCI_WIFI_ATTACH '{'
#define OPENBCI_WIFI_REMOVE '}'
#define OPENBCI_WIFI_STATUS ':'
#define OPENBCI_WIFI_RESET ';'

#endif


OpenBCI_Ganglion_Library/Honor_A_Loved_One.cpp
//==========================================================//
//     OpenBCI Kickstarter Pledge - Honor A Loved One       //
//==========================================================//

/*
Name Of Loved One To Be Honored -- Note About Honored Loved One
C.R. Zimmerman & Family -- all the best
Leonidas Hofmann -- Always bravely carrying on despite Tourette's and OCD. You rock!
Silvia Escamilla -- Te admiro, por tu voluntad y por tu fuerza, por tu cariño y por tu alegría.
Jillian Burns -- Gone to Young, Forever in our hearts.
Grandma Wang -- My grandmother suffered a stroke in 2005 and this is dedicated to her.
Kyle Scott Dickinson -- My awesome autistic brother, I'm also his personal care attendant at the north shore arc. -- He’s a great brother in fact he won't cry from pain yet only when I'm upset lol.
Robert S. Patton (my Dad) -- My Dad's life was ended by Alzheimer's. Fortunately for him, the disease took its toll quickly, less than 5 years. For him, and so many others I know, knew, and will know I encourage the founders to continue their work.
Ruth Turner -- You were always loved and are missed
Mark Wiegand -- At some point in time and space, we make you walk again.
Eleonora Pawlowska -- In memory of my beloved mother. I wish she was still here with all of us.
Dr Frederick Fries -- RIP 2010. Love you dad.
Leibniz -- Nothing is accomplished all at once, and it is one of my great maxims, and one of the most completely verified, that Nature makes no leaps: a maxim which I have called the law of continuity.
Ubirakitan Maciel -- A person dedicated to make Neuroscience accessible to people
A. Sterling -- My father survived a brain aneurysm 16 years ago He just turned 90!
Broome Spiro -- Different in an excellent sort of way.
Lauren S. -- Your depression doesn't need to define you
Stephanie Y. -- Hope you found your dream.
My dad -- Essential tremor
Peter Beals -- artist, the kindest man I ever know
*/


OpenBCI_Ganglion_Library.h
/*
   insert header here

*/
#ifndef _____OpenBCI_Ganglion_Library__
#define _____OpenBCI_Ganglion_Library__


#include <SPI.h>
#include <SimbleeBLE.h>
#include <OpenBCI_Wifi_Master.h>
#include <OpenBCI_Wifi_Master_Definitions.h>
#include "Definitions_Ganglion.h"

class OpenBCI_Ganglion {
public:


  typedef enum PACKET_TYPE {
    PACKET_TYPE_ACCEL,
    PACKET_TYPE_RAW_AUX,
    PACKET_TYPE_USER_DEFINED,
    PACKET_TYPE_ACCEL_TIME_SET,
    PACKET_TYPE_ACCEL_TIME_SYNC,
    PACKET_TYPE_RAW_AUX_TIME_SET,
    PACKET_TYPE_RAW_AUX_TIME_SYNC,
    PACKET_TYPE_IMPEDANCE
  };

  typedef enum SAMPLE_RATE {
    SAMPLE_RATE_25600,
    SAMPLE_RATE_12800,
    SAMPLE_RATE_6400,
    SAMPLE_RATE_3200,
    SAMPLE_RATE_1600,
    SAMPLE_RATE_800,
    SAMPLE_RATE_400,
    SAMPLE_RATE_200
  };

  OpenBCI_Ganglion();

  SAMPLE_RATE curSampleRate;

  void    processIncomingSampleRate(char);
  void    setSampleRate(uint8_t);
  const char * getSampleRate(void);

  void initialize(void);
  void makeUniqueId(void);
  void blinkLED(void);
  void startFromScratch(unsigned long);
  void printAllRegisters_Serial();
  void processData(void);
  boolean startRunning(void);
  boolean stopRunning(void);
  // void sendChannelData(void);
  int changeChannelState_maintainRunningState(int, int);
  void initSyntheticData(void);
  void startRunningSynthetic(void);
  void incrementSyntheticChannelData(void);
  void buildRawPacket(void);
  void sendRawPacket(void);
  void compressData18(void);
  void compressData19(void);
  void sendCompressedPacket18(void);
  void sendCompressedPacket19(void);
  void resendPacket(byte);  // this trip to the past will happen in the future
  void testImpedance(void);
  void endImpedanceTest(void);
  double convertRawGanglionImpedanceToTarget(double);
  void updateDAC(word);
  void updateDAC(void);
  void zeroDAC(void);
  float get_Zvalue(int);
  word getDACzeroPosition();
  void logData_Serial(void);
  void readShuntSensor(void);
  void gotoTarget(float, float );
  void rampTest(void);
  void changeZtestForChannel(int, int);
  void updateAccelerometerData(void);
  void config_LIS2DH(void);
  void enable_LIS2DH(void);
  void disable_LIS2DH(void);
  word LIS2DH_readTemp(void);
  byte LIS2DH_read(byte);
  void LIS2DH_write(byte, byte);
  short LIS2DH_read16(byte);
  float getG(byte);
  void LIS2DH_readAllRegs_Serial();
  // int MCP_ISR(uint32_t);
  void config_MCP3912(unsigned long);
  void updateMCPdata(void);
  void MCP_sendCommand(byte, byte);
  long MCP_readRegister(void);
  void MCP_writeRegister(unsigned long);
  void MCP_turnOnChannels(void);
  void MCP_turnOffAllChannels(void);
  void MCP_readAllRegs_Serial(void);
  void MCP_printRegisterName_Serial(byte);
  void MCP_readAllRegs(void);
  void MCP_printRegisterName(byte);
  boolean eventSerial(void);
  void sendSerialBytesBlocking(void);
  void prepToSendBytes(void);
  void loadNewLine(void);
  void loadString(const char*, int, boolean);
  void loadString(const char*);
  void loadString(void);
  void loadlnString(const char*);
  void loadlnString(void);
  void loadChar(char, boolean);
  void loadHex(int, int, boolean);
  void initSerialBuffer(void);
  void loadInt(int i, boolean);
  void parseChar(char);
  void parseCharWifi(char);
  void printFailure(void);
  void printSampleRate(void);
  void printSuccess(void);
  uint8_t * getGains(void);

  boolean settingSampleRate = false;

  uint8_t advdata[15] =
  {
    14,    // length // 0
    0x09,  // complete local name type // 1
    0x47,  // 'G' // 2
    0x61,  // 'a' // 3
    0x6E,  // 'n' // 4
    0x67,  // 'g' // 5
    0x6C,  // 'l' // 6
    0x69,  // 'i' // 7
    0x6F,  // 'o' // 8
    0x6E,  // 'n' // 9
    0x2D,  // '-' // 10
    0x54,  // 'T' // 11
    0x41,  // 'A' // 12
    0x43,  // 'C' // 13
    0x4f,  // 'O' // 14
  };

  // int LED_delayTime = 300;
  unsigned int LED_timer = 0;
  boolean LED_state = true;
  boolean is_running = false;
  boolean streamSynthetic = false;
  boolean serialBytesToSend = false;
  char serialBuffer[SERIAL_BUFFER_LENGTH][SERIAL_BUFFER_LENGTH];
  int serialIndex[SERIAL_BUFFER_LENGTH];
  unsigned long timeLastPacketSent; // used to time sending verbose BLE packets
  int bufferLevel;
  int bufferLevelCounter;
  boolean wasRunningWhenCalled = false;
  boolean useAux = false;
  boolean newAuxData = false;
  char auxData[3];
  char inChar;
  boolean gotSerial = false;
  boolean commandFromSPI = false;

  //  >>>>  LIS2DH STUFF  <<<<

  boolean useAccel = false;
  int8_t axisData[3]; // holds accelerometer x,y,z
  boolean accelOnEdge = false;
  boolean accelOffEdge = false;
  byte ID;  // holds LIS2DH device ID
  float scale_factor_gs_per_count = 0.016; // assume +/-4g, normal mode. 8bits left justified
  volatile boolean LIS_dataReady = false;
  boolean newAccelData = false;


  //  >>>>  MCP3912 STUFF <<<<
  byte compression_ring[RING_SIZE][MAX_BYTES_PER_PACKET];
  int ringBufferLevel = 0;    // const int compressionMask = 0xFFFFFFF8;
  int channelData[4];            // holds channel data
  int lastChannelData[4];       // holds last channel data
  byte rawChannelData[24];
  volatile byte sampleCounter = 0xFF;    // sample counter, for real
  volatile boolean MCP_dataReady = false;
  volatile boolean zeroth = false;
  volatile boolean first = false;


  unsigned int sampleNumber;
  unsigned long channelMask = 0x00000000;  // used to turn on selected channels
  unsigned long channelEnable[4] = {ENABLE_0, ENABLE_1, ENABLE_2, ENABLE_3};
  unsigned long channelDisable[4] = {DISABLE_0, DISABLE_1, DISABLE_2, DISABLE_3};
  byte channelAddress[4] = {CHAN_0,CHAN_1,CHAN_2,CHAN_3};
  unsigned long gain = GAIN_1;
  // unsigned long sps = SAMPLE_200;
  boolean requestToStartRunning = false;
  unsigned long thisTime;
  unsigned long thatTime;
  unsigned long thisStampTime;
  unsigned long thatStampTime;
  int timeDifference;
  boolean rising[4] = {true,true,true,true};
  unsigned long regVal;


  //  >>>>  IMPEDANCE TESTING STUFF <<<<

  boolean testingImpedance = false;
  int channelUnderZtest;
  float uAmp_Value = 0.0;           // value of measured current
  float ADC_volts_per_count = 3.0/1023.0;    //0.00293255  // 3/((2^10)-1)
  // float DAC_uVolts_per_count = 3.0/4095.0;  // .0007236;  // (3/((2^12)-1))
  float DAC_volts_per_count = 3.0/4095.0;    //0.0007236;  // 3/((2^12)-1)
  float noise = 0.1;
  short DAC_position;   // 12bit DAC resolution (4096) ~0.8mV DAC, ~5nA tXCS
  float DAC_voltage;    // DAC position converted to volts
  short DACmidline = 2047;
  int Ohms;
  short DACmask;   // used in update to add control bits
  boolean increment = true;
  boolean impedanceTesting = false;
  int impedanceSwitch[5] = {Z_TEST_1, Z_TEST_2, Z_TEST_3, Z_TEST_4, Z_TEST_REF};
  int currentChannelSetting;
  int leadOffSetting;
  int currentChannelZeroPosition;
  int currentChannelPlusTenPosition;
  int currentChannelMinusTenPosition;
  boolean rampTesting = false;
  boolean Z_noiseTesting = false;
  boolean plusTen = false;
  boolean minusTen = false;
  boolean zero = false;
  unsigned long testTimer;
  int testTime = 10000;          // 10 second test time to collect data
  unsigned long sampleTimer;
  int rampSampleTime = 100;      // 100 millisecond time between samples (10Hz)
  int noiseSampleTime = 10000;   // 10000 microsecond time between samples (100Hz)
  int gotoSampleTime = 2000;     //
  int numCurrentMeasurements = 4;
  int currentMeasurementCounter;

  //  AC WAVEFORM STUPH
  boolean ACwaveTest = false;
  unsigned long halfPeriodTimer;    // used to time each cycle of the AC wave

  word realZeroPosition = 2048;     // have to discover the '0Amps' point later
  boolean ACrising = true;            // start the square wave going up
  int Z_testTime = 1000;            // test for a second?
  unsigned long Z_testTimer;        //

  unsigned long uAsampleTimer;
  int halfWave = 100;
  int uAsampleCounter;
  int currentCounts;
  int maxPosCurrentCounts;
  int minNegCurrentCounts;
  int currentCountsAtZeroAmps;
  int peakCurrentCounts;
  int positiveRunningTotal;
  int negativeRunningTotal;
  int positiveSampleCounter;
  int negativeSampleCounter;
  boolean edge = false;
  int increased, decreased, steady;
  int negativeMean, positiveMean;
  int positiveEdge, negativeEdge, negativeEdgeCounter, positiveEdgeCounter;

  //  <<<< BLE STUFF >>>>
  char radioBuffer[20];
  char resendBuffer[20];
  boolean BLEconnected = false;
  char BLEchar[20];
  int BLEcharHead = 0;
  int BLEcharTail = 0;
  unsigned long rssiTimer;
  boolean requestForOTAenable = false;
  boolean clearForOTA = false;
  boolean writingToSD = false;
  boolean useSerial = false; // used for testing stuph
  int syntheticFrequency = 500;

};

extern OpenBCI_Ganglion ganglion;

#endif
